engine/notifications.js

```diff
@@ -0,0 +1,298 @@
+import redis from '../engine/redis-helper.js';
+import webPush from 'web-push';
+
+const ttl = 2419200;
+
+if (!process.env.GCM_API_KEY) {
+  console.warn('Set the GCM_API_KEY environment variable to support GCM');
+}
+webPush.setGCMAPIKey(process.env.GCM_API_KEY);
+
+// there is only one client (useful especially while testing)
+let client;
+function setClient(dbNumber) {
+  if (client) {
+    return Promise.resolve(client);
+  }
+  return redis.getClient(dbNumber)
+  .then(redisClient => {
+    client = redisClient;
+    return client;
+  });
+}
+
+// eported only for test as we will not need to kill database
+function quitClient(dbNumber) {
+  return setClient(dbNumber)
+  .then(() => {
+    const p = redis.quit(client);
+    client = null;
+    return p;
+  });
+}
+
+// check if such device exists in database, reject if not
+function checkDeviceId(deviceId, dbNumber) {
+  return setClient(dbNumber)
+  .then(() => redis.hgetall(client, `device-${deviceId}`))
+  .then(device => {
+    if (!device || !device.endpoint) {
+      throw new Error('Not Found');
+    }
+  });
+}
+
+// get all registrations for a deviceId
+function getRegisteredFeatures(deviceId, dbNumber) {
+  return checkDeviceId(deviceId, dbNumber)
+  .then(() => redis.smembers(client, `${deviceId}-notifications`));
+}
+
+/*
+ * send confirmation even if deviceId is already deleted
+ */
+function sendConfirmation(endpoint, key, deviceId, features, dbNumber) {
+  let payload = { title: 'Registration changed' };
+  payload.body = 'You\'re not registered to any feature';
+  if (features && features.length > 0) {
+    const numberOfFeatures = features.length;
+    const message = 'Registered to';
+    if (features.indexOf('all') >= 0) {
+      payload.body = `${message} all features`;
+    } else if (numberOfFeatures === 1) {
+      if (features.indexOf('new') === 0) {
+        payload.body = `${message} new features only`;
+      } else {
+        payload.body = `${message} one feature`;
+      }
+    } else {
+      payload.body = `${message} ${numberOfFeatures} features`;
+    }
+  }
+  payload = JSON.stringify(payload);
+  return setClient(dbNumber)
+  .then(() => {
+    if (endpoint.indexOf('https://android.googleapis.com/gcm/send') === 0) {
+      return redis.set(client, `${deviceId}-payload`, payload)
+      .then(webPush.sendNotification(endpoint, ttl));
+    }
+    return webPush.sendNotification(endpoint, ttl, key, payload);
```

Comment:
In the new Web Push encryption standard (Firefox 46+ and Chrome 50+), we need both the key and the auth secret.